src/mongocxx/options/index.hpp

```diff
@@ -0,0 +1,365 @@
+    ///
+    /// @param name
+    ///   The name of the index.
+    ///
+    void name(std::string name);
+
+    ///
+    /// The current name setting.
+    ///
+    /// @return The current name.
+    ///
+    const stdx::optional<std::string>& name() const;
+
+    ///
+    /// Whether or not to create a sparse index. Sparse indexes only reference documents with the
+    /// indexed fields.
+    ///
+    /// @param sparse
+    ///   Whether or not to create a sparse index.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-sparse/
+    ///
+    void sparse(bool sparse);
+
+    ///
+    /// The current sparse setting.
+    ///
+    /// @return The current sparse setting.
+    ///
+    const stdx::optional<bool>& sparse() const;
+
+    ///
+    /// Optionally used only in MongoDB 3.0.0 and higher. Specifies the storage engine options for
+    /// the index.
+    ///
+    /// @warning
+    ///  The base_storage_options instance must outlive this options::index object.
+    ///
+    /// @param storage_options
+    ///   The storage engine options for the index.
+    ///
+    void storage_options(base_storage_options* storage_options);
+
+    ///
+    /// The current storage_options setting.
+    ///
+    /// @return The current storage_options.
+    ///
+    const base_storage_options* storage_options() const;
+
+    ///
+    /// Set a value, in seconds, as a TTL to control how long MongoDB retains documents in this
+    /// collection.
+    ///
+    /// @param expire_after_seconds
+    ///   The amount of time, in seconds, to retain documents.
+    ///
+    /// @see https://docs.mongodb.org/manual/core/index-ttl/
+    ///
+    void expire_after_seconds(std::int32_t expire_after_seconds);
+
+    ///
+    /// The current expire_after_seconds setting.
+    ///
+    /// @return The current expire_after_seconds.
+    ///
+    const stdx::optional<std::int32_t>& expire_after_seconds() const;
+
+    ///
+    /// Sets the index version.
+    ///
+    /// @param v
+    ///   The index version.
+    ///
+    void v(std::int32_t v);
+
+    ///
+    /// The current index version.
+    ///
+    /// @return The current index version.
+    ///
+    const stdx::optional<std::int32_t>& v() const;
+
+    ///
+    /// For text indexes, sets the weight document. The weight document contains field and weight
+    /// pairs.
+    ///
+    /// @param weights
+    ///   The weight document for text indexes.
+    ///
+    void weights(bsoncxx::document::view weights);
+
+    ///
+    /// The current weights setting.
+    ///
+    /// @return The current weights.
+    ///
+    const stdx::optional<bsoncxx::document::view>& weights() const;
+
+    ///
+    /// For text indexes, the language that determines the list of stop words and the rules for the
+    /// stemmer and tokenizer.
+    ///
+    /// @param default_language
+    ///   The default language used when creating text indexes.
+    ///
+    void default_language(std::string default_language);
```

Comment:
As far as I can tell, all other options classes take these as `std::string`'s and not views. I think the members for options classes should all be owning. It's nice to be able do something like `opts.name("name")` and not have to do `auto name = "name"; opts.name(name);`.